drivers/timer/arcv2_timer0.c

```diff
@@ -512,76 +178,69 @@ int z_clock_driver_init(struct device *device)
 	return 0;
 }
 
-#ifdef CONFIG_DEVICE_POWER_MANAGEMENT
-static int sys_clock_suspend(struct device *dev)
+void z_clock_set_timeout(s32_t ticks, bool idle)
 {
-	ARG_UNUSED(dev);
+	/* If the kernel allows us to miss tick announcements in idle,
+	 * then shut off
+	 * the counter. (Note: we can assume if idle==true that
+	 * interrupts are already disabled)
+	 */
+	if (IS_ENABLED(CONFIG_TICKLESS_IDLE) && idle && ticks == K_FOREVER) {
+		timer0_control_register_set(0);
+		timer0_count_register_set(0);
+		timer0_limit_register_set(0);
+		last_load = TIMER_STOPPED;
+		return;
+	}
 
-	saved_limit = timer0_limit_register_get();
-	saved_control = timer0_control_register_get();
+#if defined(CONFIG_TICKLESS_KERNEL) && !defined(CONFIG_QEMU_TICKLESS_WORKAROUND)
+	u32_t delay;
 
-	arcv2_timer0_device_power_state = DEVICE_PM_SUSPEND_STATE;
+	ticks = min(MAX_TICKS, max(ticks - 1, 0));
 
-	return 0;
-}
+	/* Desired delay in the future */
+	delay = (ticks == 0) ? MIN_DELAY : ticks * CYC_PER_TICK;
 
-static int sys_clock_resume(struct device *dev)
-{
-	ARG_UNUSED(dev);
+	k_spinlock_key_t key = k_spin_lock(&lock);
 
-	timer0_limit_register_set(saved_limit);
-	timer0_control_register_set(saved_control);
+	delay += elapsed();
 
-	/*
-	 * It is difficult to accurately know the time spent in DS.
-	 * Expire the timer to get the scheduler called.
-	 */
-	timer0_count_register_set(saved_limit - 1);
+	/* Round delay up to next tick boundary */
+	delay = ((delay + CYC_PER_TICK - 1) / CYC_PER_TICK) * CYC_PER_TICK;
+	last_load = delay;
 
-	arcv2_timer0_device_power_state = DEVICE_PM_ACTIVE_STATE;
+	timer0_limit_register_set(last_load - 1);
+	timer0_control_register_set(_ARC_V2_TMR_CTRL_NH | _ARC_V2_TMR_CTRL_IE);
 
-	return 0;
+	k_spin_unlock(&lock, key);
+#endif
 }
 
-/*
- * Implements the driver control management functionality
- * the *context may include IN data or/and OUT data
- */
-int z_clock_device_ctrl(struct device *port, u32_t ctrl_command,
-			  void *context)
+
+
+u32_t z_clock_elapsed(void)
 {
-	if (ctrl_command == DEVICE_PM_SET_POWER_STATE) {
-		if (*((u32_t *)context) == DEVICE_PM_SUSPEND_STATE) {
-			return sys_clock_suspend(port);
-		} else if (*((u32_t *)context) == DEVICE_PM_ACTIVE_STATE) {
-			return sys_clock_resume(port);
-		}
-	} else if (ctrl_command == DEVICE_PM_GET_POWER_STATE) {
-		*((u32_t *)context) = arcv2_timer0_device_power_state;
+	if (!TICKLESS) {
```

Comment:
As above.  Can be IS_ENABLED(CONFIG_TICKLESS_KERNEL) here, no need for QEMU silliness.

Reply:
We are woring on adding arc support in QMEU,  so it may be needed in the future.

Reply:
It's a monstrostrous hack, needed by only a few tests right now.  Let's not propagate it if we aren't sure we need it.  By the time arc support lands in our SDK qemu we will hopefully have icount working.

Reply:
ok, i will remove it